magnum-cats-effect/src/main/scala/com/augustnagro/magnum/magcats/Transactor.scala

```diff
@@ -0,0 +1,158 @@
+package com.augustnagro.magnum.magcats
+
+import cats.*
+import cats.effect.kernel.*
+import cats.effect.kernel.Outcome.*
+import cats.effect.std.*
+import cats.effect.syntax.all.*
+import cats.syntax.all.*
+import com.augustnagro.magnum.*
+
+import java.sql.Connection
+import javax.sql.DataSource
+import scala.util.control.NonFatal
+
+class Transactor[F[_]: Sync] private (
+    dataSource: DataSource,
+    sqlLogger: SqlLogger,
+    connectionConfig: Connection => Unit,
+    semaphore: Option[Semaphore[F]]
+):
+  def withSqlLogger(sqlLogger: SqlLogger): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def withConnectionConfig(
+      connectionConfig: Connection => Unit
+  ): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def connect[A](f: DbCon ?=> A): F[A] =
+    val io =
+      Resource
+        .make(acquireConnection)(releaseConnection)
+        .use: cn =>
+          Sync[F].delay(connectionConfig(cn)) >>
+            Sync[F].interruptible(f(using DbCon(cn, sqlLogger)))
+
+    semaphore.fold(io)(sem => sem.permit.use { _ => io })
+
+  def transact[A](f: DbTx ?=> A): F[A] =
+    val io =
+      Resource
+        .make(acquireConnection)(releaseConnection)
+        .use: cn =>
+          Sync[F].delay(connectionConfig(cn)) >>
+            Sync[F].delay(cn.setAutoCommit(false)) >>
+            Sync[F]
+              .interruptible(f(using DbTx(cn, sqlLogger)))
+              .guaranteeCase {
+                case Succeeded(_) => Sync[F].blocking(cn.commit())
+                case Errored(_) | Canceled() =>
+                  Sync[F].blocking(cn.rollback())
+              }
+
+    semaphore.fold(io)(sem => sem.permit.use { _ => io })
+
+  private def acquireConnection: F[Connection] =
+    Sync[F]
+      .delay(dataSource.getConnection())
+      .adaptError(t => SqlException("Unable to acquire DB Connection", t))
+
+  private def releaseConnection(conn: Connection): F[Unit] =
```

Comment:
Ditto `val`

Reply:
I'm not sure which part you mean here? The `releaseConnection`?

Reply:
Oh my bad, this was in autopilot.